spec.md

```diff
@@ -3149,6 +3167,184 @@ The CO MUST implement the specified error recovery behavior when it encounters t
+Tuples in a sequence are in ascending order of `byte_offset`, with no overlap
+between adjacent tuples.
+
+```protobuf
+
+// BlockMetadata specifies a data range.
+message BlockMetadata {
+  // This is the 0 based byte position in the volume or snapshot,
+  // measured from the start of the object.
+  uint64 byte_offset = 1;
+
+  // This is the size of the data range.
+  uint64 size_bytes = 2;
+}
+
+```
+
+The `BlockMetadataType` enumerated type describes how the `size_bytes` fields of `BlockMetadata`
+tuples may vary in a sequence. There are two prevalent styles in which data ranges are described:
+
+- The **block-based** style requires a fixed value for the `size_bytes` field in all tuples in a given sequence.
+- The **extent-based** style does not constrain the value of the `size_bytes` field in a sequence.
+
+The SnapshotMetadata service permits either style at the discretion of the
+plugin as long as the style does not change mid-stream in any given RPC.
+
+```protobuf
+enum BlockMetadataType {
+  // The FIXED_LENGTH value indicates that data ranges are
+  // returned in fixed size blocks.
+  FIXED_LENGTH=0;
+
+  // The VARIABLE_LENGTH value indicates that data ranges
+  // are returned in potentially variable sized extents.
+  VARIABLE_LENGTH=1;
+}
+
+```
+
+
+#### GetAllocated RPC
+
+The plugin must implement this RPC if it provides the SnapshotMetadata service.
+
+```protobuf
+// The GetAllocatedRequest message is used to solicit metadata on the
+// allocated blocks of a snapshot: i.e. this identifies the data ranges
+// that have valid data as they were the target of some previous write
+// operation on the volume.
+message GetAllocatedRequest {
+  // This is the identifier of the snapshot.
+  string snapshot_id = 1;
+
+  // This indicates the zero based starting byte position in the volume
+  // snapshot from which the result should be computed. It is intended
+  // to be used to continue a previously interrupted call.
+  // The plugins may round down this offset to the nearest alignment
+  // boundary based on the BlockMetadataType used.
+  uint64 starting_offset = 2;
+
+  // This is an optional parameter, and if non-zero it specifies the
+  // maximum number of tuples to be returned in each
+  // GetAllocatedResponse message returned by the RPC stream.
+  // The plugin will determine an appropriate value if 0, and is
+  // always free to send less than the requested value.
+  uint32 max_results = 3;
+}
+
+// GetAllocatedResponse messages are returned in gRPC stream.
+// Cumulatively, they provide information on the allocated data
+// ranges in the snapshot.
+message GetAllocatedResponse {
+  // This specifies the style used in the BlockMetadata sequence.
+  // This value must be the same in all such messages returned by
+  // the stream.
+  BlockMetadataType block_metadata_type = 1;
+
+  // This returns the size of the underlying volume.
+  // This value must be the same in all such messages returned by
+  // the stream.
```

Comment:
The intent here is to return the "logical" volume size. I guess it would be the value returned in the `Volume` property of the `CreateVolumeResponse`:

> message Volume {
  // The capacity of the volume in bytes. This field is OPTIONAL. If not
  // set (value of 0), it indicates that the capacity of the volume is
  // unknown (e.g., NFS share).
  // The value of this field MUST NOT be negative.
  int64 capacity_bytes = 1;

Reply:
Ok, consider calling it `capacity_bytes` as well maybe?

Reply:
Okay, I'll change it to `volume_capacity_bytes` to match the `Volume` message.

Interestingly,  in `Snapshot` you'll find:
>message Snapshot {
  // This is the complete size of the snapshot in bytes. The purpose of
  // this field is to give CO guidance on how much space is needed to
  // create a volume from this snapshot. The size of the volume MUST NOT
  // be less than the size of the source snapshot. This field is
  // OPTIONAL. If this field is not set, it indicates that this size is
  // unknown. The value of this field MUST NOT be negative and a size of
  // zero means it is unspecified.
  int64 size_bytes = 1;